andela/ah-the-jedi-backend
# #165305271 Users should be able favorite and unfavorite articles.

#### What does this PR do?
Enable a user to favorite or unfavorite an article.
#### Description of Task to be completed?
- Implement `POST` and `DELETE` endpoints for favoriting and unfavoriting article respectively.
- Write tests for the implementation.
#### How should this be manually tested?
1. Clone this repo.
   `$ git clone https://github.com/andela/ah-the-jedi-backend.git`
2. Fetch all origin to get all branches
   `$ git fetch origin`
3. Checkout to `ft-user-favorite-article-165305271` branch.
   `$ git checkout ft-user-favorite-article-165305271`
4. Activate `virtualenv`
    `$ virtualenv -p python .venv`
    `$ source .venv/bin/activate`
5. Install requirements
   `$ pip install -r requirements.txt`
6. Start server
   `$ python manage.py runserver`
7. Test the endpoint on postman in the following steps:
- Ensure the user is authenticated.
- To favorite an article - `POST /api/articles/<slug>/favorite/`
- To unfavorite an article - `DEL /api/articles/<slug>/favorite/`
- Get the article you favorited or unfavorited - `GET /api/articles/<slug>/` and check for the values of `favorited` and `favoritesCount`.
#### What are the relevant pivotal tracker stories?
[#165305271](https://www.pivotaltracker.com/story/show/#165305271)
#### Screenshot
Get article before logged in, `favorited` should be `false`
<img width="1100" alt="Screenshot 2019-05-07 at 15 41 33" src="https://user-images.githubusercontent.com/17095461/57300104-c3b54c00-70de-11e9-8788-b8443eb0e0ab.png">
Get article after logged in, if its not favorited, it should be false
<img width="1074" alt="Screenshot 2019-05-07 at 15 45 57" src="https://user-images.githubusercontent.com/17095461/57300383-6241ad00-70df-11e9-81ea-c4ab21f7a970.png">
Successful favorite on an article.
<img width="1092" alt="Screenshot 2019-05-07 at 15 48 09" src="https://user-images.githubusercontent.com/17095461/57300487-974dff80-70df-11e9-8a3c-3014928704ad.png">
Get article after favoriting, `favorited` should be `true`
<img width="1086" alt="Screenshot 2019-05-07 at 15 49 46" src="https://user-images.githubusercontent.com/17095461/57300596-d1b79c80-70df-11e9-9b70-7d67db19498a.png">
Successful unfavorite on a favorited article.
<img width="1106" alt="Screenshot 2019-05-07 at 15 51 11" src="https://user-images.githubusercontent.com/17095461/57300722-12afb100-70e0-11e9-972f-e6a904129acd.png">
<img width="1097" alt="Screenshot 2019-05-07 at 15 51 34" src="https://user-images.githubusercontent.com/17095461/57301026-b9944d00-70e0-11e9-934a-7d23e5c1de9c.png">
A user cannot favorite his/her own article.
<img width="1100" alt="Screenshot 2019-05-07 at 15 53 41" src="https://user-images.githubusercontent.com/17095461/57300910-72a65780-70e0-11e9-90d4-6a956805f040.png">
